src/components/TextEditor/components/TextEditor.js

```diff
@@ -29,136 +29,152 @@ if (typeof window !== 'undefined' && typeof window.navigator !== 'undefined') {
+    saveText,
+  } = props;
 
   //= =============React Lifecycle Functions===================//
 
-  componentDidMount() {
-    window.addEventListener('beforeunload', this.onLeave);
-    window.addEventListener('close', this.onLeave);
-  }
+  useEffect(() => {
+    window.addEventListener('beforeunload', onLeave);
+    window.addEventListener('close', onLeave);
 
-  componentWillUnmount = () => {
-    window.removeEventListener('beforeunload', this.onLeave);
-    window.removeEventListener('close', this.onLeave);
-  };
+    return () => {
+      window.removeEventListener('beforeunload', onLeave);
+      window.removeEventListener('close', onLeave);
+    };
+  }, []);
 
-  openForkModal = () => {
-    this.setState({ showForkModal: true });
+  const openForkModal = () => {
+    setShowForkModal(true);
   };
 
-  closeForkModal = () => {
-    this.setState({ showForkModal: false });
+  const closeForkModal = () => {
+    setShowForkModal(false);
   };
 
-  checkDirty = async () => {
-    if (!this.props.dirty) {
+  const checkDirty = async () => {
+    if (!dirty) {
       return;
     }
 
     try {
       const programToUpdate = {};
-      programToUpdate[this.props.mostRecentProgram] = {
-        code: this.props.code,
+      programToUpdate[mostRecentProgram] = {
+        code: code,
       };
 
-      await fetch.updatePrograms(this.props.uid, programToUpdate);
+      await fetch.updatePrograms(uid, programToUpdate);
       // TODO: add functionality to be able to tell whether the fetch failed
     } catch (err) {
       console.log(err);
     }
   };
 
-  onLeave = async (ev) => {
-    if (this.props.dirty) {
+  const onLeave = async (ev) => {
+    if (dirty) {
       ev.returnValue = '';
     }
     return ev;
   };
 
-  setCodeMirrorInstance = (codeMirrorInstance) => {
-    this.setState({ codeMirrorInstance });
-  };
+  // fix this
 
-  updateCode = (editor, data, newCode) => {
+  const updateCode = (editor, data, newCode) => {
     // if the code's not yet dirty, and the old code is different from the new code, make it dirty
-    if (!this.props.dirty && this.props.code !== newCode) {
-      this.props.dirtyCode(this.props.mostRecentProgram);
+    if (!dirty && code !== newCode) {
+      dirtyCode(mostRecentProgram);
     }
-    this.props.setProgramCode(this.props.mostRecentProgram, newCode);
+    setProgramCode(mostRecentProgram, newCode);
   };
 
-  setCurrentLine = (cm) => {
-    const { codeMirrorInstance, currentLine } = this.state;
+  const setCurrentLineManual = (cm) => {
     const { line } = cm.getCursor();
     if (codeMirrorInstance) {
       // removeLineClass removes the back highlight style from the last selected line
       codeMirrorInstance.removeLineClass(currentLine, 'wrap', 'selected-line');
       // addLineClass adds the style to the newly selected line
       codeMirrorInstance.addLineClass(line, 'wrap', 'selected-line');
     }
-    this.setState({ currentLine: line });
+    setCurrentLineManual(line);
```

Comment:
function calls itself?

Reply:
whoops, will fix both these issues rn

Reply:
hmm I'm a bit confused, it says I'm up to date when I do git pull and yet the code I have isn't the same as the master.

Reply:
git merge master

Reply:
I think something is definitely wrong now

Reply:
Ok I have reverted the changes to before the merge, but every time I try to git merge master, it changes 50 files which I'm slightly concerned about.

Reply:
Honestly, I’d just take the git diff and apply it to another branch thats been freshly branched off of master 

It probably isn’t worth your time to fix up!